File: Bitwise AND of Numbers Range - Leetcode 201/Bitwise AND of Numbers Range - Leetcode 201.js
var rangeBitwiseAnd = function(m, n) {
    let shift = 0;

    // Shift both numbers to the right until they are equal
    while (m < n) {
        m >>= 1;
        n >>= 1;
        shift++;
    }

    // Shift the common prefix back to its original position
    return m << shift;
};

File: Bitwise AND of Numbers Range - Leetcode 201/Bitwise AND of Numbers Range - Leetcode 201.py
class Solution:
    def rangeBitwiseAnd(self, m: int, n: int) -> int:
        shift = 0
        
        while m < n:
            m = m >> 1
            n = n >> 1
            shift += 1

        return m << shift
        # Time: O(Bits)
        # Space: O(1)